File: 110. Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root or (not root.left and not root.right):
            return True
        if abs(self.height(root.left, 0) - self.height(root.right, 0)) > 1:
            return False
        elif (not self.isBalanced(root.left)) or (not self.isBalanced(root.right)):
            return False
        return True
        
    def height(self, root, h):
        if not root:
            return h
        return max(self.height(root.left, h), self.height(root.right, h)) + 1

File: 110. Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == NULL || (root->left == NULL && root->right == NULL)) return true;
        if (abs(height(root->left, 0) - height(root->right, 0)) > 1) return false;
        else if (!isBalanced(root->left) || !isBalanced(root->right)) return false;
        return true;
    }
    
    int height(TreeNode* root, int h)
    {
        if (root == NULL) return h;
        int l = height(root->left, h), r = height(root->right, h);
        return l > r ? l + 1 : r + 1;
    }
};
